client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,117 @@
+  static targets = ['label'];
+  static values = {
+    active: { type: String, default: gettext('Loading') },
+    durationSeconds: { type: Number, default: 30 },
+  };
+
+  activeClass: string;
+  activeValue: string;
+  durationSecondsValue: number;
+  labelTarget: HTMLElement;
+  hasActiveClass: boolean;
+  hasLabelTarget: boolean;
+
+  static afterLoad(identifier: string, application: Application) {
+    const { controllerAttribute } = application.schema;
+    const { actionAttribute } = application.schema;
+
+    const updateLoader = () => {
+      const em = document.querySelector('em') as HTMLElement;
+      document
+        .querySelectorAll(
+          `.button-longrunning:not([data-controller~='${identifier}'])`,
+        )
+        .forEach((button) => {
+          button.setAttribute(controllerAttribute, identifier);
+          button.setAttribute(actionAttribute, 'w-loader#activate');
+
+          if (button.getAttribute('data-clicked-text')) {
+            button.setAttribute(
+              'data-w-loader-active-value',
+              `${button.getAttribute('data-clicked-text')}`,
+            );
+            button.removeAttribute('data-clicked-text');
+          }
+
+          if (em.matches('button em')) {
+            em.setAttribute('data-w-loader-target', 'label');
+          }
+
+          button.setAttribute('data-w-loader-duration-seconds-value', '40');
+        });
+    };
+
+    document.addEventListener('DOMContentLoaded', updateLoader);
+  }
+
+  activate() {
+    const originalText = this.hasLabelTarget
+      ? this.labelTarget.innerText
+      : (this.element.textContent as string);
+
+    this.cancelSpinner(originalText);
+
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    // set time out
+    this.setTimeout();
+  }
+
+  cancelSpinner(originalText: string) {
+    this.element.removeAttribute('disabled');
+    this.element.classList.remove(
+      this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+    );
+
+    if (this.activeValue && this.hasLabelTarget) {
+      this.labelTarget.innerText = originalText;
+    }
+  }
+
+  setTimeout() {
+    const timeout = setTimeout(() => {
+      if (this.activeValue && this.hasLabelTarget === true) {
+        // Sets the targetElement innerText to the activeValue
+        this.labelTarget.innerText = this.activeValue;
+      }
+      // Disabling button must be done last: disabled buttons can't be
+      // modified in the normal way, it would seem.
+      this.element.classList.add(
+        this.hasActiveClass ? this.activeClass : 'button-longrunning-active',
+      );
+      this.element.setAttribute('disabled', '');
+
+      clearTimeout(timeout);
```

Comment:
We’ll additionally need to set another timeout to clear the spinner. Something like:

```suggestion
      setTimeout(() => {
        this.cancelSpinner(this.originalText);
      }, this.durationSecondsValue * 1000);
      
      clearTimeout(timeout);
```

And at the top of `activate`, line 67, store `originalText` as `this.originalText` so we can use it here and reset to the correct value.